posix/unix.c

```diff
@@ -746,9 +746,16 @@ static ssize_t send(unsigned socket, const void *buf, size_t len, int flags, con
 				if (s->type == SOCK_STREAM) {
 					err = _cbuffer_write(&conn->buffer, buf, len);
 				}
-				else if (_cbuffer_free(&conn->buffer) >= len + sizeof(len)) { /* SOCK_DGRAM or SOCK_SEQPACKET */
-					_cbuffer_write(&conn->buffer, &len, sizeof(len));
-					_cbuffer_write(&conn->buffer, buf, err = len);
+				else if (s->type == SOCK_DGRAM || s->type == SOCK_SEQPACKET) {
+					if (_cbuffer_free(&conn->buffer) >= len + sizeof(len)) {
+						_cbuffer_write(&conn->buffer, &len, sizeof(len));
+						_cbuffer_write(&conn->buffer, buf, err = len);
+					}
+					else {
+						err = -EMSGSIZE;
```

Comment:
I think your should return EMSGSIZE only if packet cannot fit to empty buffer. Currently this change breaks nonblocking send (it can return EMSGSIZE instead of EAGAIN). Something like this:
if (s->type == SOCK_STREAM) {
...
else if (_cbuffer_free(&conn->buffer) >= len + sizeof(len)) {
...
else if (conn->buffsz < len + sizeof(len)) {
  err = -EMSGSIZE;
}